source/common/common/inline_map.h

```diff
@@ -0,0 +1,460 @@
+ * key/value pairs.
+ */
+template <class StorageKey, class StorageValue> class InlineMap : public InlineStorage {
+public:
+  using TypedInlineMapDescriptor = InlineMapDescriptor<StorageKey>;
+  using Handle = typename TypedInlineMapDescriptor::Handle;
+
+  using DynamicHashMap = absl::flat_hash_map<StorageKey, StorageValue>;
+  using StorageValueRef = std::reference_wrapper<StorageValue>;
+
+  /**
+   * Get the entry by the given key. Heterogeneous lookup is supported here.
+   * @param key the key to lookup.
+   * @return the entry if the key is found, otherwise return null reference.
+   */
+  template <class LookupKey> OptRef<const StorageValue> lookup(const LookupKey& key) const {
+    // Because the dynamic key is used here, try the normal map entry first.
+    if (auto it = dynamic_entries_.find(key); it != dynamic_entries_.end()) {
+      return it->second;
+    }
+
+    if (auto entry_id = inlineLookup(key); entry_id.has_value()) {
+      if (inlineEntryValid(*entry_id)) {
+        return inline_entries_[*entry_id];
+      }
+    }
+
+    return {};
+  }
+
+  /**
+   * Get the entry by the given key. Heterogeneous lookup is supported here.
+   * @param key the key to lookup.
+   * @return the entry if the key is found, otherwise return null reference.
+   */
+  template <class LookupKey> OptRef<StorageValue> lookup(const LookupKey& key) {
+    // Because the normal string view key is used here, try the normal map entry first.
+    if (auto it = dynamic_entries_.find(key); it != dynamic_entries_.end()) {
+      return it->second;
+    }
+
+    if (auto entry_id = inlineLookup(key); entry_id.has_value()) {
+      if (inlineEntryValid(*entry_id)) {
+        return inline_entries_[*entry_id];
+      }
+    }
+
+    return {};
+  }
+
+  /**
+   * Get the entry by the given handle.
+   * @param handle the handle to lookup.
+   * @return the entry if the handle is valid, otherwise return null reference.
+   */
+  OptRef<const StorageValue> lookup(Handle handle) const {
+    if (inlineEntryValid(handle.inlineId())) {
+      return inline_entries_[handle.inlineId()];
+    }
+    return {};
+  }
+
+  /**
+   * Get the entry by the given handle.
+   * @param handle the handle to lookup.
+   * @return the entry if the handle is valid, otherwise return null reference.
+   */
+  OptRef<StorageValue> lookup(Handle handle) {
+    if (inlineEntryValid(handle.inlineId())) {
+      return inline_entries_[handle.inlineId()];
+    }
+    return {};
+  }
+
+  /**
+   * Set the entry by the given key.
+   * @param key the key to set.
+   * @param value the value to set.
+   * @return pair consisting of a reference to the inserted element, or the already-existing element
+   * if no insertion happened, and a bool denoting whether the insertion took place (true if
+   * insertion happened, false if it did not).
+   */
+  template <class LookupKey, class Value>
+  std::pair<StorageValueRef, bool> insert(LookupKey&& key, Value&& value) {
```

Comment:
consider looking to HeaderMap for inspiration here where there are separate variants for adding via move vs adding via copy.

Reply:
Hmmm, I think it's a little complex and mostly because the type difference of `HeaderString` (for full move) and LowerCaseString (for copy or ref).
I guess we needn't that here? 🤔